python/mlc_llm/bench/replay.py

```diff
@@ -0,0 +1,184 @@
+        )
+        df["timestamp"] = pd.to_datetime(df["Date"])
+        df.sort_values("timestamp", inplace=True)
+        # Get the request params from the loaded CSV
+        params = []
+        for _, row in df.iterrows():
+            request = row["@request"]
+            payload = json.loads(str(request))
+            params.append(
+                {
+                    "timestamp": row["timestamp"],
+                    "payload": payload,
+                }
+            )
+        return params
+
+    def _load_jsonl(self, filepath: str) -> List[Dict]:
+        """
+        Loads parameters from a JSONL file and returns a list of event data dictionaries.
+
+        Parameters
+        ----------
+        filepath : str
+            The path to the JSONL file.
+
+        Returns
+        -------
+        res: List[Dict]
+            A list of event data dictionaries from the JSONL file, sorted by timestamp.
+        """
+        with open(filepath, "r", encoding="utf-8") as file:
+            data = [json.loads(line) for line in file]
+            for item in data:
+                item["timestamp"] = datetime.fromisoformat(item["timestamp"])
+        data.sort(key=lambda x: x["timestamp"])
+        return data
+
+    async def send_request(self, session, params: Dict):
+        """
+        Sends an asynchronous HTTP POST request using an aiohttp session.
+
+        Parameters
+        ----------
+        session : aiohttp.ClientSession
+            The active aiohttp client session.
+
+        params : dict
+            The parameters for the request, including URL, headers, and payload, etc.
+
+        Returns
+        -------
+        res
+            The JSON response from the server or None if an error occurs.
+        """
+        try:
+            url = params.get("address", self.url)
+            headers = params.get("headers", self.headers)
+            payload = params["payload"]
+            async with session.post(
+                url, json=payload, headers=headers, timeout=self.timeout
+            ) as response:
+                return await response.json()
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            logger.error("Error in send request %s", err)
+
+    async def run(self):
+        """
+        Replays the stored requests.
+
+        Utilizes the asyncio event loop to schedule each request according to its timestamp,
+        adjusting for the current time.
+        """
+        if not self.request_params:
+            return
+        async with aiohttp.ClientSession() as session:
+            loop = asyncio.get_running_loop()
+            start_time = loop.time()
+            first_timestamp = self.request_params[0]["timestamp"]
+
+            for params in self.request_params:
+                original_delay = (params["timestamp"] - first_timestamp).total_seconds()
+                loop.call_at(
+                    start_time + original_delay,
```

Comment:
NOTE, we cannot call everything all at once. For large log, this can cause issue, have a pointer that collects `already_scheduled`. Every schedule step should have a max_delay_to_schedule, schedule all functions up to max_delay_to_schedule, then have another schedule function to be called periodically